core/src/retransmit_stage.rs

```diff
@@ -36,13 +37,99 @@ use std::{
 // it doesn't pull up too much work.
 const MAX_PACKET_BATCH_SIZE: usize = 100;
 
+#[derive(Default)]
+struct RetransmitStats {
+    total_packets: AtomicU64,
+    total_batches: AtomicU64,
+    total_time: AtomicU64,
+    repair_total: AtomicU64,
+    discard_total: AtomicU64,
+    retransmit_total: AtomicU64,
+    last_ts: AtomicU64,
+    compute_turbine_peers_total: AtomicU64,
+}
+
+fn update_retransmit_stats(
+    stats: &Arc<RetransmitStats>,
+    total_time: u64,
+    total_packets: usize,
+    retransmit_total: u64,
+    discard_total: u64,
+    repair_total: u64,
+    compute_turbine_peers_total: u64,
+    peers_len: usize,
+) {
+    stats.total_time.fetch_add(total_time, Ordering::Relaxed);
```

Comment:
To be completely accurate here I think we should change `stats.last_ts` to a `RwLock`. Then all the retransmit threads grab a read lock before they call `update_retransmit_stats` and then only have the first retransmit thread grab a write lock when the time > 1s to publish the datapoint.

This ensures that:
1) Each thread's updates across all the fields in the `stats` here either occur or don't
2) The lock should introduce a fence so that these relaxed operations don't get reordered in a way where some of these fields are very out of sync with others

Reply:
It's just stats, doesn't need to be that accurate. It's collecting from 2 seconds, an atomic swap takes 1ns, so 10 atomic swaps are not going to be skewed in time that much and the overhead of a lock is a lot more.